Fix generalized dice computation

surajpaib opened the pull request:
Fixes #7966 

### Description

A few sentences describing the changes proposed in this pull request.

### Types of changes
<!--- Put an `x` in all the boxes that apply, and remove the not applicable items -->
- [x] Non-breaking change (fix or new feature that would not break existing functionality).
- [ ] Breaking change (fix or new feature that would cause existing functionality to change).
- [ ] New tests added to cover the changes.
- [ ] Integration tests passed locally by running `./runtests.sh -f -u --net --coverage`.
- [ ] Quick tests passed locally by running `./runtests.sh --quick --unittests  --disttests`.
- [ ] In-line docstrings updated.
- [ ] Documentation updated, tested `make html` command in the `docs/` folder.

KumoLiu commented on monai/metrics/generalized_dice.py:
```diff
@@ -20,109 +20,108 @@
+    y: torch.Tensor,
+    include_background: bool = True,
+    weight_type: Weight | str = Weight.SQUARE,
+    sum_over_labels: bool = False,
```
I would suggest we set this arg as `True` by default to be consistent with previous behaviour.

surajpaib replied:
@KumoLiu 

After looking a bit more at the `meandice` implementation as well, its probably best to make both consistent. 


```python
import torch
from monai.metrics import compute_generalized_dice, compute_dice

a = torch.randn((20, 10, 64, 128, 128))
b = torch.ones((20, 10, 64, 128, 128))



compute_generalized_dice(a, b).shape, compute_dice(a, b).shape
```

The current output looks like so:
```
(torch.Size([20]), torch.Size([20, 10]))
```

KumoLiu replied:
@surajpaib, Thanks for the details, agreed, please just go ahead.

KumoLiu commented on monai/metrics/generalized_dice.py:
```diff
@@ -20,109 +20,108 @@
+            sum_over_labels=self.sum_over_labels,
         )
 
-    def aggregate(self, reduction: MetricReduction | str | None = None) -> torch.Tensor:
```
Is there any specific reason remove the reduction here? It seems that the keeping this argument maintains the same logic as other metrics.

surajpaib replied:
I've removed the reduction here because the value of `sum_over_labels` needs to be set during the `_compute_tensor` and the tensor will already be reduced over labels at that point. 

So I'm not sure how we can handle a different reduction passed to `aggregate` right now - unless we store an intermediary tensor for the `intersection` and `denominator` in the buffer and make the computation in aggregate. 

Do you have any suggestions here?

surajpaib replied:
I agree that the inconsistency might be an issue. We could offer lesser options for reduction as a solution - providing only `mean_batch` and `mean` reductions. 

This would mean you can't compute GeneralizedDice per class but that is probably fine as per the original formulation in the paper which reduces over classes by default.

surajpaib replied:
Was originally inspired by this implementation: https://lightning.ai/docs/torchmetrics/stable/segmentation/generalized_dice.html

which does provide values per class but the `per_class` flag is defined during init

surajpaib replied:
@KumoLiu Do you think we can follow the `deprecated_arg_default` convention here as well for the reduction on aggregate?

KumoLiu replied:
Sure, we can do it here. Just ensure we can get the exactly same result with before and let users aware if any behaviour will be changed.

KumoLiu replied:
Hi @surajpaib, I may misunderstand before, seems we don't need to deprecated the reduction in aggregation. sum_over_classes will only sum over class, then if users also want calculate batch, what should they specify?

surajpaib replied:
The `sum_over_classes` is only so we can enable computing gen.dice considering all classes in the numerator and denominator - the default setting for gen dice as in the original paper. 

So if the user wants to reduce `mean_batch` then the `sum_over_classes` will be set to False. But if they want to compute `mean_channel` or overall `mean`, this should be set to True as it will consider all classes in the computation. 

P.S: I realize that I need to add `mean_channel` for setting the `sum_over_classes` flag as well. Will do so asap

KumoLiu commented on monai/metrics/generalized_dice.py:
```diff
@@ -20,109 +20,108 @@
-        weight_type (Union[Weight, str], optional): {``"square"``, ``"simple"``, ``"uniform"``}. Type of function to
+        weight_type: {``"square"``, ``"simple"``, ``"uniform"``}. Type of function to
             transform ground truth volume into a weight factor. Defaults to ``"square"``.
+        sum_over_labels: Whether to sum the numerator and denominator across all labels before the final computation.
```
Do you thinks sum_over_classes would be better? Just a suggestion, open for this one.
```suggestion
        sum_over_labels: Whether to sum the numerator and denominator across all classes before the final computation.
```

surajpaib replied:
Agreed! Will make the change

KumoLiu reviewed (commented):
Thanks for the contribution, overall looks good to me.
Once you finished the PR, please let me know, I will add more people help review it, thanks.

ericspod commented on monai/metrics/generalized_dice.py:
```diff
@@ -20,109 +20,108 @@
-        include_background: bool = True,
-        reduction: MetricReduction | str = MetricReduction.MEAN_BATCH,
-        weight_type: Weight | str = Weight.SQUARE,
+        self, include_background: bool = True, reduction: str = "mean", weight_type: Weight | str = Weight.SQUARE
```
The `reduction` default value should not be a string literal (this is like a magic number essentially) and should use the value from the `MetricReduction` enum. The original default was also mean_batch and not mean, this shouldn't be changed or if it must there should be a deprecated warning using `deprecated_arg_default`.

surajpaib replied:
I'll make the changes for the enum. 

For the default reduction, the current value specified is `mean_batch` but the calling aggregate actually returns the `mean` reduction. 

This is because the `compute_generalized_dice` function reduces over classes already so applying a mean batch after it results in an overall mean.

KumoLiu replied:
Like @ericspod suggested, it would also be better to add a `deprecated_arg_default`, and mentioned that the default result of `mean_batch` before is not correct, if user want to achieve before results, they should use `mean`. Does this make sense?

KumoLiu commented:
Hi @surajpaib, do you plan to address the comments on the PR? I’m hoping we can get this merged into version 1.4, but we only have about one or two weeks left to add new features.

surajpaib commented:
Hi @KumoLiu, sorry for the gap, was away on vacation. I think I can work on this in the next few days and get it in.

surajpaib commented:
@KumoLiu I've incorporated the comments we discussed. 

`deprecated_arg_default` is added to the initialization and for the reduction, I've added `deprecated_arg`. 

The tests seem to be failing because of the version string - I assume we can set this appropriately during the release. 

Let me know if you see any other changes

KumoLiu commented on monai/metrics/generalized_dice.py:
```diff
@@ -14,115 +14,131 @@
+        "reduction",
+        old_default=MetricReduction.MEAN_BATCH,
+        new_default=MetricReduction.MEAN,
+        since="1.3.3",
+        replaced="1.3.3",
+        msg_suffix="Old versions computed `mean` when `mean_batch` was provided due to bug in reduction",
```
Instead of immediately updating the default behavior, we need to provide users with some time to adapt to this change. We should also issue a warning to inform users that the mean_batch behavior is no longer the same as before, since it previously calculated the mean. If users want to retain the previous behavior, they will need to explicitly set the parameter to "mean".
```suggestion
        since="1.4.0",
        replaced="1.5.0",
        msg_suffix="Old versions computed `mean` when `mean_batch` was provided due to bug in reduction， If you want to retain the old behavior (calculating the mean), please explicitly set the parameter to 'mean'.",
```

cc @ericspod

KumoLiu commented on monai/metrics/generalized_dice.py:
```diff
@@ -14,115 +14,131 @@
         self,
         include_background: bool = True,
-        reduction: MetricReduction | str = MetricReduction.MEAN_BATCH,
+        reduction: MetricReduction | str = MetricReduction.MEAN,
```
```suggestion
        reduction: MetricReduction | str = MetricReduction.MEAN_BATCH,
```

surajpaib commented on monai/metrics/generalized_dice.py:
```diff
@@ -50,79 +60,85 @@ def __init__(
-            raise ValueError(f"reduction must be one of {reduction_options}")
+        self.reduction = look_up_option(reduction, MetricReduction)
         self.weight_type = look_up_option(weight_type, Weight)
+        self.sum_over_classes = self.reduction == MetricReduction.SUM or self.reduction == MetricReduction.MEAN
```
@KumoLiu Just added it! Missed it on the previous, added a `mean_channel` computation in the test as well

KumoLiu commented on tests/test_compute_generalized_dice.py:
```diff
@@ -22,7 +22,7 @@
 _device = "cuda:0" if torch.cuda.is_available() else "cpu"
 
 # keep background
-TEST_CASE_1 = [  # y (1, 1, 2, 2), y_pred (1, 1, 2, 2), expected out (1)
```
Please also update the shape for Line 31, [0.8] --> [[0.8]]

KumoLiu reviewed (approved):
Thanks for the update, LGTM.
Could you please help fix the ci issue and dco issue then I can help trigger more tests and merge this one?
https://github.com/Project-MONAI/MONAI/actions/runs/10733660015/job/29767399315?pr=7970#step:7:428

KumoLiu commented:
Hi @surajpaib, you can fix the dco issue by following the method here: https://github.com/Project-MONAI/MONAI/pull/7970/checks?check_run_id=29787940944
Also please update here: https://github.com/Project-MONAI/MONAI/pull/7970/files#r1746799238
Then I will try to merge this one, thanks!

KumoLiu commented:
/build
